rpc/transport: Release resource units right after dispatching send

bharathv opened the pull request:
## Cover letter

Currently we have a timer that throws in case of timeouts but does not actually cleanup the underlying resource units. So if the dispatch of send blocks due to a remote issue (eg: in flush), the resources are held up until that returns or fails. In the linked issue, we believe the sequence of events is as follows.

* raft `_op_lock` is grabbed by replicate_entries_stm and propagated to the transport#send (append_entries)
* Test issues a shut down (blocks on consensus stop, request in progress)
* send blocked for 2+ mins holding the op_lock resources (to another node in a questionable state)
* Test timed out after 60 secs
* send returned, request failed and the shutdown got unblocked.

Looking closely, we can actually release the units early as dispatch of send guarantees the ordering for us by using a semaphore.

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes #5324

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [ ] not a bug fix
- [ ] papercut/not impactful enough to backport
- [x] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

* None

## Release notes

* none

### Features

* None.

### Improvements

* Recovers from failures quickly by cleaning up resources.

bharathv commented on src/v/rpc/test/rpc_gen_cycling_test.cc:
```diff
@@ -337,15 +342,87 @@ FIXTURE_TEST(timeout_test, rpc_integration_fixture) {
+
+// Make sure all the resources are released, including the ones that have run
+// into dispatch issues (if any).
+tests::cooperative_spin_wait_with_timeout(5s, [&] {
```
I messed up something here in git-clang-format.. let me force push and fix this. (something got overwritten).

bharathv commented on src/v/rpc/test/rpc_gen_cycling_test.cc:
```diff
@@ -337,15 +342,86 @@ FIXTURE_TEST(timeout_test, rpc_integration_fixture) {
+    std::vector<::mutex> mutexes;
+    mutexes.reserve(num_sends);
+
+    // A bunch of RPCs with large sleep, this creates congestion.
```
The test doesn't run into the problem deterministically (I only saw it happen a few times of many runs).. I  retained it because it does some sort of load tests with delayed RPCs..open to new ideas / removing this if seems unnecessary.

mmaslankaprv replied:
Maybe it would be good to write a ducktape test where one of the nodes is isolated and we try to restart nodes that tries to communicate with the isolated one ?

bharathv replied:
I added one.

bharathv commented:
I'm testing it locally with induced sleeps in the flush code path  + a load_generator in the background. Then I'm SIGINT-ing  brokers in a rolling fashion and so far I can see brokers shutdown ok.. I'll keep running this test for a few hours,  analyse the logs and report here.
Only issue with these timeouts is that they are creating a lot of noise in other RPCs causing unnecessary leadership transfers and voting rounds (and a ton of uncaught exceptions traces in the logs).

mmaslankaprv commented on src/v/rpc/test/rpc_gen_cycling_test.cc:
```diff
@@ -347,6 +350,36 @@ FIXTURE_TEST(timeout_test, rpc_integration_fixture) {
+    auto opts = rpc::client_opts(rpc::clock_type::now() + 1s);
+    opts.resource_units = ss::make_foreign(
+      ss::make_lw_shared(std::move(units)));
+    ;
```
not needed semicolon

mmaslankaprv commented on src/v/rpc/transport.cc:
```diff
@@ -192,33 +193,41 @@ transport::do_send(sequence_t seq, netbuf b, rpc::client_opts opts) {
+                    _requests_queue.erase(it->first);
+                    auto f = _out.write(std::move(v));
+                    return ss::with_timeout(t, std::move(f))
+                      .handle_exception([](const std::exception_ptr& e) {
```
I think that here we should use `handle_exception_type` and handle only timeout exceptions, now each timeout will cause all the outstanding requests to fail.

mmaslankaprv replied:
you are right

mmaslankaprv commented on src/v/rpc/transport.cc:
```diff
@@ -215,7 +215,7 @@ void transport::dispatch_send(const client_opts& opts) {
                     return ss::with_timeout(t, std::move(f))
                       .handle_exception([](const std::exception_ptr& e) {
-                          vlog(rpclog.debug, "Error dispatching send() {}", e);
+                          vlog(rpclog.info, "Error dispatching send() {}", e);
```
unrelated change

mmaslankaprv commented on src/v/rpc/transport.cc:
```diff
@@ -192,33 +193,41 @@ transport::do_send(sequence_t seq, netbuf b, rpc::client_opts opts) {
-            _probe.request_error();
-            fail_outstanding_futures();
-        });
+      = ssx::spawn_with_gate_then(
```
can you split this commit into two where: 
1) would apply change to the dispatch logic
2) would change formatting

mmaslankaprv replied:
oh, sorry i think i wasn't clear, i meant this `dispatch_send` method, it doesn't really change but the indentation does making it hard to track actual changes.

bharathv replied:
Oh no worries, now that I removed the timeout handling, I don't think indents is an issue anymore? Also I guess I could just retain the coroutinization change, seems like the direction we generally want to move in for code hygeine?

bharathv replied:
@mmaslankaprv Now that we have a release branch, I updated the PR with your suggestion..The ordering test that we chatted about is exactly the same one in rpc_gen_cycling_test::`ordering_test`, so didn't add a new one.

bharathv reviewed (commented):
Added test fails in buildkite, need to take a closer look

bharathv commented:
/ci-repeat 5

bharathv commented:
/ci-repeat 5

bharathv commented:
The issue was with not cleaning up failure injector (ISOLATE) properly on teardown. I updated the patch and `ci-repeat-5` looks ok, only failure in the run was `VerifiableProducer` not stopping in time but that is not related to the core fix here.. I'll look into it but the patch is ready for review.

mmaslankaprv commented on src/v/rpc/transport.cc:
```diff
@@ -182,33 +183,37 @@ transport::do_send(sequence_t seq, netbuf b, rpc::client_opts opts) {
+                    _requests_queue.erase(it->first);
+                    auto f = _out.write(std::move(v));
+                    return ss::with_timeout(t, std::move(f))
+                      .finally([this, msg_size, units = std::move(units)] {
```
i am now wondering if we really need to keep those units alive until the write is finished, we use `batched_output_stream` it use a semaphore when processing `write`. The semaphore has a waiters queue that guaranteed ordering. what do you think about simply not capturing `units` in finally continuation ?

mmaslankaprv replied:
i mentioned it here as it seems that we wouldn't need a timeout handling then

mmaslankaprv commented on src/v/rpc/transport.cc:
```diff
@@ -19,6 +19,7 @@
 #include "vlog.h"
 
 #include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
```
not needed include

mmaslankaprv commented:
@dotnwat can you please take a look at this, it looks good to me

dotnwat commented on src/v/rpc/test/rpc_gen_cycling_test.cc:
```diff
@@ -363,6 +364,35 @@ FIXTURE_TEST(timeout_test, rpc_integration_fixture) {
+    BOOST_REQUIRE(lock.try_get_units().has_value() == false);
+    auto echo_resp = client.sleep_for(
+      echo::sleep_req{.secs = 10}, std::move(opts));
+    BOOST_REQUIRE_EQUAL(
+      echo_resp.get0().error(), rpc::errc::client_request_timeout);
+    // Verify that the resource is released before the RPC finishes.
+    tests::cooperative_spin_wait_with_timeout(5s, [&] {
+        return lock.try_get_units().has_value();
+    }).get();
```
how does this verify that the resource is related before the RPC finishes? `echo_resp.get0()` blocks until a response is received, and only after that, do we do a test of the lock.

bharathv replied:
you're right, that comment is incorrect.. I'll fix it (just to be clear.. the intent was to add coverage for release of units after RPC finishes with a timeout.).

dotnwat commented on src/v/rpc/transport.cc:
```diff
@@ -169,44 +169,33 @@ transport::make_response_handler(netbuf& b, const rpc::client_opts& opts) {
+        maybe_update_last_seq(seq);
+        co_return errc::disconnected_endpoint;
+    }
+    auto f = make_response_handler(b, opts);
```
i'm concerned that `make_response_handler` is called after `co_await get_units()`. before this change to use coroutines, `make_response_handler` would have run uninterrupted in the same context as the calling fiber. are there some implicit behaviors that depend on that running uninterrupted? probably not, but it does seem like we should preserve the original semantics unless we we don't want to preserve them in which case we can look more closely at what is happening. wdyt?

bharathv replied:
I agree with you, I actually forgot why I made this change.. vaguely remember it was intentional but I can't remember why. But now that I look at it again, I think the new version has an issue.. making the response handler prior to grabbing mem units means that .. timeout on the RPC can kick in (and clean up caller state) if there is memory pressure and we can't get a hold of the units within the timeout window. Original version looks correct, wdyt?

dotnwat replied:
> Original version looks correct, wdyt?

do you mean the version prior to coroutinize such that it needs to be fixed (or the commit dropped)?

dotnwat replied:
i wonder if that commit should be dropped? it doesn't look related to the reosurce units change, right?

dotnwat reviewed (commented):
looks good, but i think we should consider preserving the original concurrency semantics in the changes to `transport::do_send(sequence_t seq, netbuf b, rpc::client_opts opts) {`, not because what you have doesn't look correct, but because the rpc bits are so fragile and subtle it's hard to analyze correctness w.r.t. some implicit ordering requirements. wdyt?

bharathv commented:
/ci-repeat 5

bharathv commented:
/ci-repeat 10

bharathv commented:
> looks good, but i think we should consider preserving the original concurrency semantics in the changes to transport::do_send(sequence_t seq, netbuf b, rpc::client_opts opts) {, not because what you have doesn't look correct, but because the rpc bits are so fragile and subtle it's hard to analyze correctness w.r.t. some implicit ordering requirements. wdyt?

@dotnwat Took a deeper look at the CRC mismatch issue, so here is the summary

* It is _not_ related to coroutinization of do_send, reproduces without that commit.
* It is _not_ related to the failure injection in the test, reproduces without it.
* It is _not_ related to the current patch at all, reproduces on the tip of `dev`
* It _is_ related to the write throughput in the producer, I reduced it and the errors are gone. I've dug into the `VerifiableProducer` that wraps a regular kafka client, nothing jumps out. Does not repro for me locally at all (even in 1000s of runs). Seems like a client side issue (I hope) but I don't have any theory.

That said, I removed the coroutinization commit just to be extra safe and fixed the test by reducing the throughput, mind taking another look?

bharathv commented:
/ci-repeat 10

dotnwat commented:
> It is not related to the current patch at all, reproduces on the tip of dev
> It is related to the write throughput in the producer, I reduced it and the errors are gone. I've dug into the VerifiableProducer that wraps a regular kafka client, nothing jumps out. Does not repro for me locally at all (even in 1000s of runs). Seems like a client side issue (I hope) but I don't have any theory.

@bharathv is there a ticket tracking the failure and how to reproduce it?

bharathv commented:
> @bharathv is there a ticket tracking the failure and how to reproduce it?

Tracking in https://github.com/redpanda-data/redpanda/issues/6631.

BenPope commented:
I updated the cover letter with backport info, since I think I've seen it here: https://buildkite.com/redpanda/redpanda/builds/16044#01839d62-8fd9-420d-ad3c-2e3f7edb4af9 - please adjust if that's not the case.

dotnwat reviewed (approved):
lgtm.

@bharathv in some later PR it might be useful to sneak in a comment on batch output stream that the rpc layer depends on ordering imposed by the write interface. for example, it should never wake-up multiple waiters on the semaphore because those may lose their original FIFO ordering.

bharathv commented:
> in some later PR it might be useful to sneak in a comment on batch output stream that the rpc layer depends on ordering imposed by the write interface. for example, it should never wake-up multiple waiters on the semaphore because those may lose their original FIFO ordering.

That makes sense.

bharathv commented:
/backport 22.2.x

dotnwat commented:
@bharathv should this not be backported? IIUC it isn't fixing a specific bug

bharathv commented:
> @bharathv should this not be backported? IIUC it isn't fixing a specific bug

Original theory was that this fixes the shutdown hang reported in #5324. And Ben recently reported an [occurrence](https://github.com/redpanda-data/redpanda/pull/5730#issuecomment-1268312138) in 22.2.x branch, hence the backport. You think it is risky for a patch release?
